teamscale_client/precommit_client.py

```diff
@@ -0,0 +1,160 @@
+
+
+precommit_config_filename = '.teamscale-precommit.config'
+
+
+def _parse_args():
+    """Parses the precommit client command line arguments."""
+    parser = argparse.ArgumentParser(description='Precommit analysis client for Teamscale.')
+    parser.add_argument('path', metavar='path', type=str, nargs=1,
+                        help='path to any file in the repository')
+    parser.add_argument('--exclude-findings-in-changed-code', dest='exclude_findings_in_changed_code',
+                        action='store_const', const=True, default=False,
+                        help='Determines whether to exclude findings in changed code (default: False)')
+    parser.add_argument('--fetch-existing-findings', dest='fetch_existing_findings', action='store_const',
+                        const=True, default=False,
+                        help='When this option is set, existing findings in the specified file are fetched in addition '
+                             'to precommit findings. (default: False)')
+    parser.add_argument('--fetch-all-findings', dest='fetch_all_findings', action='store_const',
+                        const=True, default=False,
+                        help='When this option is set, all existing findings in the repo are fetched in addition '
+                             'to precommit findings. (default: False)')
+    return parser.parse_args()
+
+
+def configure_precommit_client(config_file, repo_path, parsed_args):
+    """Reads the precommit analysis configuration and creates a precommit client with the corresponding config."""
+    parser = ConfigParser()
+    parser.read(config_file)
+    config = TeamscaleConfig(url=parser.get('teamscale', 'url'), username=parser.get('teamscale', 'username'),
+                             access_token=parser.get('teamscale', 'access_token'),
+                             project_id=parser.get('project', 'id'))
+    return PrecommitClient(teamscale_config=config, repository_path=repo_path, analyzed_file=parsed_args.path[0])
+
+
+def analyze():
+    """Performs precommit analysis."""
+    parsed_args = _parse_args()
+    repo_path = get_repo_root_from_file_in_repo(parsed_args.path[0])
+    if not repo_path or not os.path.exists(repo_path) or not os.path.isdir(repo_path):
+        raise RuntimeError('Invalid path to file in repository: %s' % repo_path)
+
+    print('Configuring precommit analysis...')
+
+    config_file = os.path.join(repo_path, precommit_config_filename)
+    if not os.path.exists(config_file) or not os.path.isfile(config_file):
+        raise RuntimeError('Config file could not be found: %s' % config_file)
+    precommit_client = configure_precommit_client(config_file=config_file, repo_path=repo_path, parsed_args=parsed_args)
+
+    print('Uploading changes for precommit analysis for repo at %s...' % repo_path)
+    precommit_client.upload_precommit_data()
+
+    # We need to wait for the analysis to pick up the new code otherwise we get old findings
+    time.sleep(2)
```

Comment:
We might not need this in the future, I think this was fixed as part of a bug recently